File: Surface_tracking_matlab/DecomrotationPro.m
function [alfa,beta,gamma] = DecomrotationPro(R)
% Contribution
%  Author: Xinyu Nie
%  Created: 2022/11/10
%  Copyright:The Neuro Image Computing Research (NICR) group at the Mark and Mary Stevens Neuroimaging 
%  and Informatics Institute of USC Laboratory of NeuroImaging 
%  USC Stevens Neuroimaging and Informatics Institute
%  email: xnie@usc.edu

xr=R(:,1);
%yr=R(:,2);
zr=R(:,3);
xq=sqrt(sum(xr.^2));
xr=xr/xq;
%yq=sqrt(sum(yr.^2));
%yr=yr/yq;
zq=sqrt(sum(zr.^2));
zr=zr/zq;
x=[1;0;0];
y=[0;1;0];
z=[0;0;1];
n=cross(z,zr);
qn=sqrt(sum(n.^2));
if  qn>0
    n=n/qn;
end
  alfa=acos(sum(n.*y));
  if n(1)>0
      alfa =2*pi - alfa;
  end
  Rz=[cos(alfa) -sin(alfa) 0; sin(alfa) cos(alfa) 0; 0 0 1];
  x1=Rz*x;
  %y1=n;
  beta=acos(sum(z.*zr));
  Rn=rotationMatrix(n,beta);
  zr1=Rn*z;
  pz1=abs(sum(zr1.*zr));
  Rnr=rotationMatrix(n,2*pi-beta);
  zr11=Rnr*z;
  pz2=abs(sum(zr11.*zr));
  if pz1 < pz2
      beta = 2*pi-beta;
  end
  x2=rotationMatrix(n,beta)*x1;
  gamma=acos(sum(x2.*xr));
  Rzr=rotationMatrix(zr,gamma);
  xr1=Rzr*x2;
  px1=abs(sum(xr1.*xr));
  Rzrr=rotationMatrix(zr,2*pi-gamma);
  xr11=Rzrr*x2;
  px2=abs(sum(xr11.*xr));
  if px1< px2
      gamma=2*pi -gamma;
  end

end

File: SurfacePro/DecomrotationPro.cpp
#include "DecomrotationPro.h"
vector<float> DecomrotationPro(float R[][3]){
    
    float* xr = R[0];
    float* zr = R[2];
    normalize(xr);
    normalize(zr);
    float x[3] = {1,0,0}, y[3] = {0,1,0}, z[3] = {0,0,1};
    float n[3];
    cross(n,z,zr);
    float qn = norm(n);
    if (qn>0) {n[0]/=qn; n[1]/=qn; n[2]/=qn;}
    float alfa = acos(dot(n,y));
    if (n[0]>0) alfa = 2*PI - alfa;
    float Rz[3][3] = {{cos(alfa), -sin(alfa), 0}, {sin(alfa), cos(alfa), 0}, {0,0,1}};
    float x1[3];
    for (int i=0;i<3;i++) x1[i] = dot(Rz[i],x);
    float beta = acos(dot(z,zr));
    float Rn[3][3];
    rotate(Rn, n, beta);
    float zr1[3];
    for (int i =0;i<3;i++) zr1[i] = dot(Rn[i],z);
    float pz1 = abs(dot(zr1,zr));
    float Rnr[3][3];
    rotate(Rnr, n,2*PI-beta);
    float zr11[3];
    for (int i = 0;i<3;i++) zr11[i] = dot(Rnr[i],z);
    float pz2 = abs(dot(zr11,zr));
    if (pz1 < pz2){
        beta = 2*PI-beta;
    }
    float tmp[3][3];
    rotate(tmp,n,beta);
    float x2[3];
    for (int i =0;i<3;i++) x2[i] = dot(tmp[i],x1);
    float gamma = acos(dot(x2,xr));
    float Rzr[3][3];
    rotate(Rzr,zr,gamma);
    float xr1[3];
    for (int i=0;i<3;i++) xr1[i] = dot(Rzr[i],x2);
    float px1 = abs(dot(xr1,xr));
    float Rzrr[3][3];
    rotate(Rzrr, zr, 2*PI-gamma);
    float xr11[3];
    for (int i =0;i<3;i++) xr11[i] = dot(Rzrr[i], x2);
    float px2 = abs(dot(xr11,xr));
    if (px1<px2) gamma = 2*PI-gamma;
    return {alfa, beta, gamma};
}